tests/capture/test_capture_while_loop.py

```diff
@@ -0,0 +1,199 @@
+
+    @pytest.mark.parametrize("x", [1.6, 2.4])
+    def test_while_loop_simple(self, x):
+        """Test simple while-loop primitive"""
+
+        def fn(x):
+
+            @qml.while_loop(lambda x: x < 2)
+            def loop(x):
+                return x**2
+
+            x2 = loop(x)
+            return x2
+
+        expected = x**2 if x < 2 else x
+        result = fn(x)
+        assert np.allclose(result, expected), f"Expected {expected}, but got {result}"
+
+        jaxpr = jax.make_jaxpr(fn)(x)
+        res_ev_jxpr = jax.core.eval_jaxpr(jaxpr.jaxpr, jaxpr.consts, x)
+        assert np.allclose(res_ev_jxpr, expected), f"Expected {expected}, but got {res_ev_jxpr}"
+
+    @pytest.mark.parametrize("array", [jax.numpy.zeros(0), jax.numpy.zeros(5)])
+    def test_while_loop_dynamic_array(self, array):
+        """Test while loops with dynamic array inputs."""
+
+        def fn(arg):
+
+            a, b = jax.numpy.ones(arg.shape, dtype=float), jax.numpy.ones(arg.shape, dtype=float)
+
+            # Note: lambda *_, idx: idx < 10 doesn't work - necessary keyword argument not provided
+            @qml.while_loop(lambda *args: args[-1] < 10)
+            def loop(a, b, idx):
+                return a + b, b + a, idx + 2
+
+            return loop(a, b, 0)
+
+        res_arr1, res_arr2, res_idx = fn(array)
+        expected = 2**5 * jax.numpy.ones(*array.shape)
+        assert jax.numpy.allclose(res_arr1, res_arr2)
+        assert jax.numpy.allclose(res_arr1, expected), f"Expected {expected}, but got {res_arr1}"
+
+        jaxpr = jax.make_jaxpr(fn)(array)
+        res_arr1_jxpr, res_arr2_jxpr, res_idx_jxpr = jax.core.eval_jaxpr(
+            jaxpr.jaxpr, jaxpr.consts, array
+        )
+
+        assert np.allclose(res_arr1_jxpr, res_arr2_jxpr)
+        assert np.allclose(res_arr1_jxpr, expected), f"Expected {expected}, but got {res_arr1_jxpr}"
+        assert np.allclose(res_idx, res_idx_jxpr) and res_idx_jxpr == 10
+
+
+class TestCaptureCircuitsWhileLoop:
+    """Tests for capturing for while loops into jaxpr in the context of quantum circuits."""
+
+    def test_while_loop_capture(self):
+        """Test that a while loop is correctly captured into a jaxpr."""
+
+        dev = qml.device("default.qubit", wires=3)
+
+        @qml.qnode(dev)
+        def circuit():
+
+            @qml.while_loop(lambda i: i < 3)
+            def loop_fn(i):
+                qml.RX(i, wires=0)
+                return i + 1
+
+            _ = loop_fn(0)
+
+            return qml.expval(qml.Z(0))
+
+        result = circuit()
+        expected = np.cos(0 + 1 + 2)
+        assert np.allclose(result, expected), f"Expected {expected}, but got {result}"
+
+        jaxpr = jax.make_jaxpr(circuit)()
+        res_ev_jxpr = jax.core.eval_jaxpr(jaxpr.jaxpr, jaxpr.consts)
+        assert np.allclose(res_ev_jxpr, expected), f"Expected {expected}, but got {res_ev_jxpr}"
+
+    @pytest.mark.parametrize("arg, expected", [(1.2, -0.16852022), (1.6, 0.598211352)])
```

Comment:
I know I let it through on the `for_loop` PR, so I won't block on this.  I have just created a story in the technical backlog to create better tests when we have time

But hard-coded magic results like this can cause a lot of problems down the line.  If we suddenly start getting different results from this test and things are failing, it can be hard to tell if we actually fixed something and the hard coded number was incorrect, or whether we are now incorrect. But determining what answers we expect from a complicated workflow is not always possible.

So concerned, but not blocking.